host_core/lib/host_core/control_interface/server.ex

```diff
@@ -143,22 +143,24 @@ defmodule HostCore.ControlInterface.Server do
   # Put a link definition
   # This will first store the link definition in memory, then publish it to the stream
   # then publish it directly to the relevant provider via the RPC channel
+  @put_linkdef_keys ["actor_id", "contract_id", "link_name", "provider_id", "values"]
   defp handle_request({"linkdefs", "put"}, body, _reply_to) do
     Tracer.with_span "Handle Linkdef Put (ctl)", kind: :server do
       with {:ok, ld} <- Jason.decode(body),
-           true <-
-             ["actor_id", "contract_id", "link_name", "provider_id", "values"]
-             |> Enum.all?(&Map.has_key?(ld, &1)) do
-        HostCore.Linkdefs.Manager.put_link_definition(
-          ld["actor_id"],
-          ld["contract_id"],
-          ld["link_name"],
-          ld["provider_id"],
-          ld["values"]
-        )
-
+           true <- has_keys?(@put_linkdef_keys, ld),
+           :ok <-
+             HostCore.Linkdefs.Manager.put_link_definition(
+               ld["actor_id"],
+               ld["contract_id"],
+               ld["link_name"],
+               ld["provider_id"],
+               ld["values"]
+             ) do
         {:reply, success_ack()}
       else
+        {:error, {:duplicate_key, {_, _, link_name}}} ->
+          {:reply, failure_ack("Duplicate link definition entry for '#{link_name}'")}
```

Comment:
I think this failure ack would be more informative as
```
Link definition for {actor_id} on contract {contract_id} and link {link_name} already exists and must be deleted first
```

Reply:
It's a little wordy but I'd rather include all the information than not enough

Reply:
Agree